metrics/tags.go

```diff
@@ -5,8 +5,83 @@ import (
 	"encoding/json"
 	"sort"
 	"strings"
+
+	"github.com/mstoykov/atlas"
 )
 
+// TagSet represents a set of tags.
+type TagSet struct {
+	tags *atlas.Node
+}
+
+// NewTagSet creates a TagSet, if a not-nil map is passed
+// then it adds all the pairs to the set, otherwise an empty set is created.
+// Under the hood it initializes an Atlas root node.
+// It should be only used in a centralized place and
+// all the new TagSet should branch from it for getting
+// the optimal performances.
+func NewTagSet(m map[string]string) *TagSet {
+	node := atlas.New()
+	for k, v := range m {
+		node = node.AddLink(k, v)
+	}
+
+	return &TagSet{
+		tags: node,
+	}
+}
+
+// TagSetFromSampleTags creates a TagSet starting
+// from the set defined by SampleTags.
+// Adding a tag on the new set doesn't impact the SampleTags.
+func TagSetFromSampleTags(st *SampleTags) *TagSet {
+	tm := &TagSet{}
+	tm.tags = st.tags
+	return tm
+}
+
+// AddTag adds a tag in the set.
+func (tg *TagSet) AddTag(k, v string) {
+	tg.tags = tg.tags.AddLink(k, v)
+}
+
+// Get returns the Tag value and true
+// if the provided key has been found.
+func (tg *TagSet) Get(k string) (string, bool) {
+	return tg.tags.ValueByKey(k)
+}
+
+// Len returns the number of the set keys.
+func (tg *TagSet) Len() int {
+	return tg.tags.Len()
+}
+
+// Delete deletes the item related to the provided key.
+func (tg *TagSet) Delete(k string) {
+	tg.tags = tg.tags.DeleteKey(k)
+}
+
+// Map returns a map of string pairs with the items in the TagSet.
+func (tg *TagSet) Map() map[string]string {
+	return tg.tags.Path()
+}
+
+// BranchOut creates a new TagSet with the same set of items
+// as in the current TagSet.
+// Any new added tag to the new set creates a new dedicated path.
+func (tg *TagSet) BranchOut() *TagSet {
+	tmcopy := &TagSet{}
+	tmcopy.tags = tg.tags
+	return tmcopy
+}
+
+// SampleTags creates a SampleTags using the current tag set.
+func (tg *TagSet) SampleTags() *SampleTags {
+	st := &SampleTags{}
+	st.tags = tg.tags
+	return st
+}
+
```

Comment:
Connected to https://github.com/grafana/k6/pull/2594/files#r937514286

Can't we have it retunr the atlas.Node and a "constructor" to basically take that :thinking: 

If we decide to drop SampleTags (or at least make it be just *atlas.Node I guess) that will be sufficent and might actually be easier to work with :thinking: 

The `branchFromRoot.....` will actually be `metrics.TagSetFromSample(registry.RootTagsNode())`

Reply:
We had a video call to resolve this and the notes are in https://github.com/grafana/k6/pull/2594#issuecomment-1205273413